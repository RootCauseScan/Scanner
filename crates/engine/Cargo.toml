[package]
name = "engine"
version = "1.0.0"
edition = "2021"
description = "Parallel rule evaluation engine for RootCause"
readme = "README.md"
license.workspace = true
repository = "https://github.com/rootcausescan/scanner"
homepage = "https://github.com/rootcausescan/scanner"
keywords = ["security", "sast", "static-analysis"]
categories = ["development-tools", "code-analysis"]

[dependencies]
rayon = { workspace = true }
regex = { workspace = true }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true }
loader = { path = "../loader" }
ir = { path = "../ir" }
rego_wasm = { path = "../rego_wasm" }
tokio = { version = "1", features = ["rt-multi-thread", "time"] }
anyhow = "1"
tracing = { workspace = true }
blake3 = "1"
parsers = { path = "../parsers" }
shlex = { workspace = true }
plugin_core = { path = "../plugin_core" }

[dev-dependencies]
parsers = { path = "../parsers" }
tempfile = "3"
tracing-subscriber = { workspace = true }
base64 = { workspace = true }
