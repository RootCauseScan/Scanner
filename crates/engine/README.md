# engine

Analysis engine that evaluates rules over the intermediate representation with parallel execution and cache.

## Build

```bash
cargo build -p engine
```

## Test

```bash
cargo test -p engine
```

## Documentation

- [Architecture details](../../documentation/architecture/crates/engine.md)

## Contributing

See [CONTRIBUTING.md](../CONTRIBUTING.md) for guidelines.
