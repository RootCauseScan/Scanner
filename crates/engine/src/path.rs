use regex::Regex;
use std::collections::{HashMap, VecDeque};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Mutex, OnceLock, RwLock};

pub static CANONICAL_PATHS: OnceLock<RwLock<HashMap<PathBuf, PathBuf>>> = OnceLock::new();
pub static CANONICAL_STATS: OnceLock<CacheStats> = OnceLock::new();
pub static CANONICAL_ORDER: OnceLock<RwLock<VecDeque<PathBuf>>> = OnceLock::new();
static PATH_REGEX_CACHE: OnceLock<Mutex<PathRegexCache>> = OnceLock::new();

#[cfg(test)]
pub const CANONICAL_CACHE_CAPACITY: usize = 3;
#[cfg(not(test))]
pub const CANONICAL_CACHE_CAPACITY: usize = 1024;

#[cfg(test)]
pub const PATH_REGEX_CACHE_CAPACITY: usize = 3;
#[cfg(not(test))]
pub const PATH_REGEX_CACHE_CAPACITY: usize = 1024;

#[derive(Default)]
pub struct CacheStats {
    hits: AtomicUsize,
    misses: AtomicUsize,
}

struct PathRegexCache {
    map: HashMap<String, Regex>,
    order: VecDeque<String>,
    capacity: usize,
}

impl PathRegexCache {
    fn new(capacity: usize) -> Self {
        Self {
            map: HashMap::new(),
            order: VecDeque::new(),
            capacity,
        }
    }

    fn get(&mut self, key: &str) -> Option<&Regex> {
        if let Some(pos) = self.order.iter().position(|k| k == key) {
            let k = self.order.remove(pos).expect("order index must exist");
            self.order.push_back(k);
            return self.map.get(key);
        }
        None
    }

    fn insert(&mut self, key: String, value: Regex) {
        if self.map.contains_key(&key) {
            if let Some(pos) = self.order.iter().position(|k| k == &key) {
                self.order.remove(pos);
            }
        }
        self.order.push_back(key.clone());
        self.map.insert(key, value);
        if self.order.len() > self.capacity {
            if let Some(oldest) = self.order.pop_front() {
                self.map.remove(&oldest);
            }
        }
    }

    #[cfg(test)]
    fn clear(&mut self) {
        self.map.clear();
        self.order.clear();
    }

    #[cfg(test)]
    fn len(&self) -> usize {
        self.map.len()
    }

    #[cfg(test)]
    fn contains(&self, key: &str) -> bool {
        self.map.contains_key(key)
    }
}

pub fn canonicalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path_ref = path.as_ref();
    let cache = CANONICAL_PATHS.get_or_init(|| RwLock::new(HashMap::new()));
    let order = CANONICAL_ORDER.get_or_init(|| RwLock::new(VecDeque::new()));
    let stats = CANONICAL_STATS.get_or_init(Default::default);
    if let Some(cached) = cache
        .read()
        .unwrap_or_else(|e| e.into_inner())
        .get(path_ref)
        .cloned()
    {
        stats.hits.fetch_add(1, Ordering::Relaxed);
        let mut ord = order.write().unwrap_or_else(|e| e.into_inner());
        if let Some(pos) = ord.iter().position(|p| p == path_ref) {
            ord.remove(pos);
        }
        ord.push_back(path_ref.to_path_buf());
        return cached;
    }
    stats.misses.fetch_add(1, Ordering::Relaxed);
    let canonical = fs::canonicalize(path_ref).unwrap_or_else(|_| {
        #[cfg(windows)]
        {
            PathBuf::from(path_ref.to_string_lossy().replace('\\', "/"))
        }
        #[cfg(not(windows))]
        {
            path_ref.to_path_buf()
        }
    });
    let mut map = cache.write().unwrap_or_else(|e| e.into_inner());
    let mut ord = order.write().unwrap_or_else(|e| e.into_inner());
    map.insert(path_ref.to_path_buf(), canonical.clone());
    ord.push_back(path_ref.to_path_buf());
    if ord.len() > CANONICAL_CACHE_CAPACITY {
        if let Some(oldest) = ord.pop_front() {
            map.remove(&oldest);
        }
    }
    canonical
}

pub(crate) fn cache_stats() -> (usize, usize) {
    let stats = CANONICAL_STATS.get_or_init(Default::default);
    (
        stats.hits.load(Ordering::Relaxed),
        stats.misses.load(Ordering::Relaxed),
    )
}

#[cfg(test)]
pub fn reset_canonical_cache() {
    if let Some(map) = CANONICAL_PATHS.get() {
        map.write().unwrap_or_else(|e| e.into_inner()).clear();
    }
    if let Some(ord) = CANONICAL_ORDER.get() {
        ord.write().unwrap_or_else(|e| e.into_inner()).clear();
    }
    if let Some(stats) = CANONICAL_STATS.get() {
        stats.hits.store(0, Ordering::Relaxed);
        stats.misses.store(0, Ordering::Relaxed);
    }
}

#[cfg(test)]
pub fn canonical_cache_stats() -> (usize, usize) {
    cache_stats()
}

#[cfg(test)]
pub fn reset_path_regex_cache() {
    if let Some(cache) = PATH_REGEX_CACHE.get() {
        cache.lock().unwrap().clear();
    }
}

#[cfg(test)]
pub fn path_regex_cache_size() -> usize {
    PATH_REGEX_CACHE
        .get()
        .map(|c| c.lock().unwrap().len())
        .unwrap_or(0)
}

#[cfg(test)]
pub fn path_regex_cache_contains(pat: &str) -> bool {
    PATH_REGEX_CACHE
        .get()
        .map(|c| c.lock().unwrap().contains(pat))
        .unwrap_or(false)
}

pub fn path_matches(pattern: &str, candidate: &str) -> bool {
    if !pattern
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || "/._-".contains(c) || c == '*')
    {
        return false;
    }
    let cache =
        PATH_REGEX_CACHE.get_or_init(|| Mutex::new(PathRegexCache::new(PATH_REGEX_CACHE_CAPACITY)));
    let mut cache = cache.lock().expect("path regex cache lock poisoned");
    if let Some(rx) = cache.get(pattern) {
        return rx.is_match(candidate);
    }
    let mut re = String::from("^");
    for ch in pattern.chars() {
        if ch == '*' {
            re.push_str(".*");
        } else {
            re.push_str(&regex::escape(&ch.to_string()));
        }
    }
    re.push('$');
    match Regex::new(&re) {
        Ok(rx) => {
            let is_match = rx.is_match(candidate);
            cache.insert(pattern.to_string(), rx);
            is_match
        }
        Err(_) => false,
    }
}
