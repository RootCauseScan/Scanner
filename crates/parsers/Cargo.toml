[package]
name = "parsers"
version = "1.0.0"
edition = "2021"
description = "Parsers that convert source files into RootCause IR"
readme = "README.md"
license.workspace = true
repository = "https://github.com/rootcausescan/scanner"
homepage = "https://github.com/rootcausescan/scanner"
keywords = ["security", "sast", "static-analysis"]
categories = ["development-tools", "code-analysis"]

[dependencies]
anyhow = "1"
rayon = { workspace = true }
regex = { workspace = true }
serde = { workspace = true,  features = ["derive"] }
serde_yaml = { workspace = true }
serde_json = { workspace = true }
ir = { path = "../ir" }
tree-sitter = "0.20"
tree-sitter-typescript = "0.20"
tree-sitter-python = "0.20"
tree-sitter-go = "0.20"
tree-sitter-ruby = "0.20"
tree-sitter-rust = "0.20"
tree-sitter-javascript = "0.20"
tree-sitter-java = "0.20"
tree-sitter-php = "0.20"
marked-yaml = "0.8"
hcl-edit = "0.9"
tracing = { workspace = true }
blake3 = "1"

[dev-dependencies]
tempfile = "3"

