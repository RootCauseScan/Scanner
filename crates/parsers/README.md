# parsers

Collection of parsers for Dockerfile, YAML, HCL and various code languages. Generates `FileIR` and, optionally, ASTs.

## Build

```bash
cargo build -p parsers
```

## Test

```bash
cargo test -p parsers
```

## Documentation

- [Supported languages](../../documentation/architecture/crates/parsers)

## Contributing

See [CONTRIBUTING.md](../CONTRIBUTING.md) for guidelines.
