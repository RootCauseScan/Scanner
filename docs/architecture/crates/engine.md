---
id: engine
title: Engine
description: Analysis engine that evaluates rules on the intermediate representation
sidebar_position: 2
---

# engine

## Responsibility
Evaluates rules on the intermediate representation of files, handling parallel execution, caching and optional WASM support.

## Key functions and structures
- `analyze_files_cached`: analyzes files applying hash-based caching.
- `build_cfg`: builds control flow graphs for flow analysis.
- `Finding`: individual result with severity, location and message.

## Usage example
```rust
use engine::{analyze_files_cached, EngineConfig};
use loader::load_rules;
use parsers::parse_file;
use std::path::Path;

let rules = load_rules(Path::new("rules")).unwrap();
let file = parse_file(Path::new("a.yaml"), None, None).unwrap().unwrap();
let findings = analyze_files_cached(&[file], &rules, Path::new("cache.json"), &EngineConfig::default(), None);
assert!(findings.is_empty());
```
