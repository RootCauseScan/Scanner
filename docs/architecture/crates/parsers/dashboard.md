---
id: dashboard
title: Dashboard
description: Parsers to convert files from different languages to intermediate representation
sidebar_position: 1
---

# Dashboard

## Key functions and structures
- `detect_type`: determines the file type from its name or extension.
- `parse_file`: reads and transforms a file into `FileIR`.

## Maturity Guide

For detailed information about parser maturity levels (L1-L8), evaluation criteria and current status of each language, see the [Maturity Guide](./maturity.md).

## Supported languages

| Language | Level | Status |
| --- | --- | --- |
| Python | L5 | Intermediate |
| Rust | L2 | Syntactic |
| Dockerfile | L1 | Prototype |
| YAML | L1 | Prototype |
| HCL (Terraform) | L1 | Prototype |
| TypeScript | L1 | Prototype |
| JavaScript | L1 | Prototype |
| Go | L1 | Prototype |
| Java | L4 | Intermediate |
| PHP | L4 | Intermediate |
| Ruby | L1 | Prototype |

*For complete details on maturity levels and evaluation criteria, see the [Maturity Guide](./maturity.md).*

## Usage example
```rust
use parsers::parse_file;
use std::fs;
use std::path::Path;
let path = std::env::temp_dir().join("ex.yaml");
fs::write(&path, "a: 1").unwrap();
let ir = parse_file(Path::new(&path), None, None).unwrap().unwrap();
assert_eq!(ir.nodes[0].path, "a");
```

## Contributing new parsers

1. Create a module in `crates/parsers/src/languages/`.
2. Export it in `crates/parsers/src/languages/mod.rs` and update `detect_type` and `parse_file` in `crates/parsers/src/lib.rs`.
3. Add fixtures in `examples/fixtures/<language>` with `good` and `bad` cases.
4. Write tests in `crates/parsers/tests` using those fixtures.
