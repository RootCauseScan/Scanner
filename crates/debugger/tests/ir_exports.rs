use engine::{build_cfg, parse_file_with_events};
use parsers::build_dfg;
use std::path::Path;

fn load_fir() -> ir::FileIR {
    parse_file_with_events(
        Path::new("../../examples/fixtures/python/py.no-eval/bad.py"),
        None,
        None,
    )
    .unwrap()
    .unwrap()
}

#[test]
fn ast_to_dot() {
    let fir = load_fir();
    let ast = fir.ast.expect("ast");
    insta::assert_snapshot!(ast.to_dot());
}

#[test]
fn ast_to_json() {
    let fir = load_fir();
    let ast = fir.ast.expect("ast");
    insta::assert_snapshot!(ast.to_json().unwrap());
}

#[test]
fn cfg_to_dot() {
    let fir = load_fir();
    let cfg = build_cfg(&fir).expect("cfg");
    insta::assert_snapshot!(cfg.to_dot());
}

#[test]
fn cfg_to_json() {
    let fir = load_fir();
    let cfg = build_cfg(&fir).expect("cfg");
    insta::assert_snapshot!(cfg.to_json().unwrap());
}

#[test]
fn ssa_to_dot() {
    let mut fir = load_fir();
    build_dfg(&mut fir).unwrap();
    let dfg = fir.dfg.expect("dfg");
    insta::assert_snapshot!(dfg.to_dot());
}

#[test]
fn ssa_to_json() {
    let mut fir = load_fir();
    build_dfg(&mut fir).unwrap();
    let dfg = fir.dfg.expect("dfg");
    insta::assert_snapshot!(dfg.to_json().unwrap());
}
