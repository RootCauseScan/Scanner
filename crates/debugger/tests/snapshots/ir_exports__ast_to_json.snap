---
source: crates/debugger/tests/ir_exports.rs
assertion_line: 26
expression: ast.to_json().unwrap()
---
{
  "file_path": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
  "file_type": "python",
  "nodes": [
    {
      "id": 0,
      "parent": null,
      "kind": "FunctionDefinition",
      "value": "unsafe",
      "children": [
        {
          "id": 1,
          "parent": 0,
          "kind": "Def",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 1
          }
        },
        {
          "id": 2,
          "parent": 0,
          "kind": "Identifier",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 5
          }
        },
        {
          "id": 3,
          "parent": 0,
          "kind": "Parameters",
          "value": null,
          "children": [
            {
              "id": 4,
              "parent": 3,
              "kind": "(",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 1,
                "column": 11
              }
            },
            {
              "id": 5,
              "parent": 3,
              "kind": "Identifier",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 1,
                "column": 12
              }
            },
            {
              "id": 6,
              "parent": 3,
              "kind": ")",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 1,
                "column": 16
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 11
          }
        },
        {
          "id": 7,
          "parent": 0,
          "kind": ":",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 17
          }
        },
        {
          "id": 8,
          "parent": 0,
          "kind": "Block",
          "value": null,
          "children": [
            {
              "id": 9,
              "parent": 8,
              "kind": "ExpressionStatement",
              "value": null,
              "children": [
                {
                  "id": 10,
                  "parent": 9,
                  "kind": "Call",
                  "value": "eval",
                  "children": [
                    {
                      "id": 11,
                      "parent": 10,
                      "kind": "Identifier",
                      "value": null,
                      "children": [],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 5
                      }
                    },
                    {
                      "id": 12,
                      "parent": 10,
                      "kind": "ArgumentList",
                      "value": null,
                      "children": [
                        {
                          "id": 13,
                          "parent": 12,
                          "kind": "(",
                          "value": null,
                          "children": [],
                          "meta": {
                            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                            "line": 2,
                            "column": 9
                          }
                        },
                        {
                          "id": 14,
                          "parent": 12,
                          "kind": "Identifier",
                          "value": null,
                          "children": [],
                          "meta": {
                            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                            "line": 2,
                            "column": 10
                          }
                        },
                        {
                          "id": 15,
                          "parent": 12,
                          "kind": ")",
                          "value": null,
                          "children": [],
                          "meta": {
                            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                            "line": 2,
                            "column": 14
                          }
                        }
                      ],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 9
                      }
                    }
                  ],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 5
                  }
                }
              ],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 5
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 5
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 1
      }
    }
  ],
  "index": [
    {
      "id": 0,
      "parent": null,
      "kind": "FunctionDefinition",
      "value": "unsafe",
      "children": [
        {
          "id": 1,
          "parent": 0,
          "kind": "Def",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 1
          }
        },
        {
          "id": 2,
          "parent": 0,
          "kind": "Identifier",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 5
          }
        },
        {
          "id": 3,
          "parent": 0,
          "kind": "Parameters",
          "value": null,
          "children": [
            {
              "id": 4,
              "parent": 3,
              "kind": "(",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 1,
                "column": 11
              }
            },
            {
              "id": 5,
              "parent": 3,
              "kind": "Identifier",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 1,
                "column": 12
              }
            },
            {
              "id": 6,
              "parent": 3,
              "kind": ")",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 1,
                "column": 16
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 11
          }
        },
        {
          "id": 7,
          "parent": 0,
          "kind": ":",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 17
          }
        },
        {
          "id": 8,
          "parent": 0,
          "kind": "Block",
          "value": null,
          "children": [
            {
              "id": 9,
              "parent": 8,
              "kind": "ExpressionStatement",
              "value": null,
              "children": [
                {
                  "id": 10,
                  "parent": 9,
                  "kind": "Call",
                  "value": "eval",
                  "children": [
                    {
                      "id": 11,
                      "parent": 10,
                      "kind": "Identifier",
                      "value": null,
                      "children": [],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 5
                      }
                    },
                    {
                      "id": 12,
                      "parent": 10,
                      "kind": "ArgumentList",
                      "value": null,
                      "children": [
                        {
                          "id": 13,
                          "parent": 12,
                          "kind": "(",
                          "value": null,
                          "children": [],
                          "meta": {
                            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                            "line": 2,
                            "column": 9
                          }
                        },
                        {
                          "id": 14,
                          "parent": 12,
                          "kind": "Identifier",
                          "value": null,
                          "children": [],
                          "meta": {
                            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                            "line": 2,
                            "column": 10
                          }
                        },
                        {
                          "id": 15,
                          "parent": 12,
                          "kind": ")",
                          "value": null,
                          "children": [],
                          "meta": {
                            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                            "line": 2,
                            "column": 14
                          }
                        }
                      ],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 9
                      }
                    }
                  ],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 5
                  }
                }
              ],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 5
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 5
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 1
      }
    },
    {
      "id": 1,
      "parent": 0,
      "kind": "Def",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 1
      }
    },
    {
      "id": 2,
      "parent": 0,
      "kind": "Identifier",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 5
      }
    },
    {
      "id": 3,
      "parent": 0,
      "kind": "Parameters",
      "value": null,
      "children": [
        {
          "id": 4,
          "parent": 3,
          "kind": "(",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 11
          }
        },
        {
          "id": 5,
          "parent": 3,
          "kind": "Identifier",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 12
          }
        },
        {
          "id": 6,
          "parent": 3,
          "kind": ")",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 1,
            "column": 16
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 11
      }
    },
    {
      "id": 4,
      "parent": 3,
      "kind": "(",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 11
      }
    },
    {
      "id": 5,
      "parent": 3,
      "kind": "Identifier",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 12
      }
    },
    {
      "id": 6,
      "parent": 3,
      "kind": ")",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 16
      }
    },
    {
      "id": 7,
      "parent": 0,
      "kind": ":",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 1,
        "column": 17
      }
    },
    {
      "id": 8,
      "parent": 0,
      "kind": "Block",
      "value": null,
      "children": [
        {
          "id": 9,
          "parent": 8,
          "kind": "ExpressionStatement",
          "value": null,
          "children": [
            {
              "id": 10,
              "parent": 9,
              "kind": "Call",
              "value": "eval",
              "children": [
                {
                  "id": 11,
                  "parent": 10,
                  "kind": "Identifier",
                  "value": null,
                  "children": [],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 5
                  }
                },
                {
                  "id": 12,
                  "parent": 10,
                  "kind": "ArgumentList",
                  "value": null,
                  "children": [
                    {
                      "id": 13,
                      "parent": 12,
                      "kind": "(",
                      "value": null,
                      "children": [],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 9
                      }
                    },
                    {
                      "id": 14,
                      "parent": 12,
                      "kind": "Identifier",
                      "value": null,
                      "children": [],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 10
                      }
                    },
                    {
                      "id": 15,
                      "parent": 12,
                      "kind": ")",
                      "value": null,
                      "children": [],
                      "meta": {
                        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                        "line": 2,
                        "column": 14
                      }
                    }
                  ],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 9
                  }
                }
              ],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 5
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 5
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 5
      }
    },
    {
      "id": 9,
      "parent": 8,
      "kind": "ExpressionStatement",
      "value": null,
      "children": [
        {
          "id": 10,
          "parent": 9,
          "kind": "Call",
          "value": "eval",
          "children": [
            {
              "id": 11,
              "parent": 10,
              "kind": "Identifier",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 5
              }
            },
            {
              "id": 12,
              "parent": 10,
              "kind": "ArgumentList",
              "value": null,
              "children": [
                {
                  "id": 13,
                  "parent": 12,
                  "kind": "(",
                  "value": null,
                  "children": [],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 9
                  }
                },
                {
                  "id": 14,
                  "parent": 12,
                  "kind": "Identifier",
                  "value": null,
                  "children": [],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 10
                  }
                },
                {
                  "id": 15,
                  "parent": 12,
                  "kind": ")",
                  "value": null,
                  "children": [],
                  "meta": {
                    "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                    "line": 2,
                    "column": 14
                  }
                }
              ],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 9
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 5
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 5
      }
    },
    {
      "id": 10,
      "parent": 9,
      "kind": "Call",
      "value": "eval",
      "children": [
        {
          "id": 11,
          "parent": 10,
          "kind": "Identifier",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 5
          }
        },
        {
          "id": 12,
          "parent": 10,
          "kind": "ArgumentList",
          "value": null,
          "children": [
            {
              "id": 13,
              "parent": 12,
              "kind": "(",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 9
              }
            },
            {
              "id": 14,
              "parent": 12,
              "kind": "Identifier",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 10
              }
            },
            {
              "id": 15,
              "parent": 12,
              "kind": ")",
              "value": null,
              "children": [],
              "meta": {
                "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
                "line": 2,
                "column": 14
              }
            }
          ],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 9
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 5
      }
    },
    {
      "id": 11,
      "parent": 10,
      "kind": "Identifier",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 5
      }
    },
    {
      "id": 12,
      "parent": 10,
      "kind": "ArgumentList",
      "value": null,
      "children": [
        {
          "id": 13,
          "parent": 12,
          "kind": "(",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 9
          }
        },
        {
          "id": 14,
          "parent": 12,
          "kind": "Identifier",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 10
          }
        },
        {
          "id": 15,
          "parent": 12,
          "kind": ")",
          "value": null,
          "children": [],
          "meta": {
            "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
            "line": 2,
            "column": 14
          }
        }
      ],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 9
      }
    },
    {
      "id": 13,
      "parent": 12,
      "kind": "(",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 9
      }
    },
    {
      "id": 14,
      "parent": 12,
      "kind": "Identifier",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 10
      }
    },
    {
      "id": 15,
      "parent": 12,
      "kind": ")",
      "value": null,
      "children": [],
      "meta": {
        "file": "/workspace/rust-sast/examples/fixtures/python/py.no-eval/bad.py",
        "line": 2,
        "column": 14
      }
    }
  ]
}
